diff --git a/tiling.js b/tiling.js
index 4e98ed7..30af917 100644
--- a/tiling.js
+++ b/tiling.js
@@ -1123,11 +1123,11 @@ export class Space extends Array {
         return true;
     }
 
-    switchLeft(loop) { return this.switch(Meta.MotionDirection.LEFT, loop); }
-    switchRight(loop) { return this.switch(Meta.MotionDirection.RIGHT, loop); }
-    switchUp(loop) { return this.switch(Meta.MotionDirection.UP, loop); }
-    switchDown(loop) { return this.switch(Meta.MotionDirection.DOWN, loop); }
-    switch(direction, loop) {
+    switchLeft(loop, focus = true) { return this.switch(Meta.MotionDirection.LEFT, loop, focus); }
+    switchRight(loop, focus = true) { return this.switch(Meta.MotionDirection.RIGHT, loop, focus); }
+    switchUp(loop, focus = true) { return this.switch(Meta.MotionDirection.UP, loop, focus); }
+    switchDown(loop, focus = true) { return this.switch(Meta.MotionDirection.DOWN, loop, focus); }
+    switch(direction, loop, focus = true) {
         let space = this;
         let index = space.selectedIndex();
         if (index === -1) {
@@ -1180,7 +1180,21 @@ export class Space extends Array {
 
         let metaWindow = space.getWindow(index, row);
         ensureViewport(metaWindow, space);
+        if (!focus)
+            return true;
 
+        // Not using ensureViewport({callback}) because the navigator will invoke minimap,
+        // and the final focus is determined via the minimap, instead of here.
+        signals.connectOneShot(Navigator.getNavigator(), 'destroy', () => {
+            // Wait the animation, otherwise the selected window will report a wrong rect:
+            const timeout = Settings.prefs.animation_time * 1000;
+            if (this.warpPointerTimeout)
+                clearTimeout(this.warpPointerTimeout);
+            this.warpPointerTimeout = setTimeout(() => {
+                Utils.warpPointerToMonitor(null, display.focus_window);
+                this.warpPointerTimeout = null;
+            }, timeout);
+        });
         return true;
     }
 
@@ -2544,9 +2558,6 @@ export const Spaces = class Spaces extends Map {
         if (i === -1)
             return;
         let newMonitor = Main.layoutManager.monitors[i];
-        if (warp) {
-            Utils.warpPointerToMonitor(newMonitor);
-        }
         let space = this.monitors.get(newMonitor);
 
         if (move && focus) {
@@ -2575,6 +2586,10 @@ export const Spaces = class Spaces extends Map {
         } else {
             space.activate(false, false);
         }
+
+        if (warp) {
+            Utils.warpPointerToMonitor(newMonitor, display.focus_window);
+        }
     }
 
     moveToMonitor(direction, backDirection) {
diff --git a/topbar.js b/topbar.js
index f42a528..2c992ca 100644
--- a/topbar.js
+++ b/topbar.js
@@ -180,10 +180,10 @@ export function topBarScrollAction(event) {
     let direction = event.get_scroll_direction();
     switch (direction) {
     case Clutter.ScrollDirection.DOWN:
-        Tiling.spaces?.activeSpace.switchRight(false);
+        Tiling.spaces?.activeSpace.switchRight(false, false);
         break;
     case Clutter.ScrollDirection.UP:
-        Tiling.spaces?.activeSpace.switchLeft(false);
+        Tiling.spaces?.activeSpace.switchLeft(false, false);
         break;
     }
     const selected = Tiling.spaces?.activeSpace?.selectedWindow;
diff --git a/utils.js b/utils.js
index 15bb86a..2ec0109 100644
--- a/utils.js
+++ b/utils.js
@@ -234,28 +234,28 @@ export function monitorAtCurrentPoint() {
 
 /**
  * Warps pointer to the center of a monitor.
+ * Nope, we're warpping pointer to the focues window.
  */
-export function warpPointerToMonitor(monitor, params = { center: false, ripple: true }) {
-    const center = params?.center ?? false;
-    const ripple = params?.ripple ?? true;
+export function warpPointerToMonitor(monitor = null, window = null) {
+    const ripple = monitor != null;
 
-    // no need to warp if already on this monitor
-    let currMonitor = monitorAtCurrentPoint();
-    if (currMonitor === monitor) {
+    // force to wrap even if we're on the same monitor:
+    if (window) {
+        const rect = window.get_buffer_rect();
+        warpPointer(
+            rect.x + rect.width / 2,
+            rect.y + rect.height / 2,
+            ripple);
         return;
     }
 
-    let [x, y] = global.get_pointer();
-    if (center) {
-        x -= monitor.x;
-        y -= monitor.y;
-        warpPointer(
-            monitor.x + Math.floor(monitor.width / 2),
-            monitor.y + Math.floor(monitor.height / 2),
-            ripple);
+    // no need to warp if already on this monitor
+    let currMonitor = monitorAtCurrentPoint();
+    if (!monitor || currMonitor === monitor) {
         return;
     }
 
+    let [x, y] = global.get_pointer();
     let proportionalX = (x - currMonitor.x) / currMonitor.width;
     let proportionalY = (y - currMonitor.y) / currMonitor.height;
     warpPointer(
