diff --git a/Makefile b/Makefile
index e362f81..dccfa40 100644
--- a/Makefile
+++ b/Makefile
@@ -41,7 +41,7 @@ scaled/info.xml: $(SCALED)
 		name="$${filename%.*}"; \
 		echo "    <wallpaper>" >> "build/info.xml"; \
 		echo "        <name>$$name</name>" >> "build/info.xml"; \
-		echo "        <filename>/usr/share/backgrounds/pop/$$(basename "$$file")</filename>" >> "build/info.xml"; \
+		echo "        <filename>/run/current-system/sw/share/backgrounds/pop/$$(basename "$$file")</filename>" >> "build/info.xml"; \
 		echo "        <options>zoom</options>" >> "build/info.xml"; \
 		echo "        <pcolor>#000000</pcolor>" >> "build/info.xml"; \
 		echo "        <scolor>#000000</scolor>" >> "build/info.xml"; \
